### PR TITLE
Implement Datetime date component

### PR DESCRIPTION
**Pull request by RFDarter**
Implements the Datetime component from this PR

Related PRs:
- https://github.com/esphome/esphome/pull/6191
- https://github.com/home-assistant/core/pull/112475

### CODE REVIEW

**Comment by bdraco** [comment]
Please generate the protobuf files with

```
          docker run \
            -v "$PWD":/aioesphomeapi \
            ghcr.io/esphome/aioesphomeapi-proto-builder:latest
```

**Comment by RFDarter** [comment]
> ghcr.io/esphome/aioesphomeapi-proto-builder

running
`docker run -v "$PWD":/aioesphome ghcr.io/esphome/aioesphomeapi-proto-builder:latest`
results in 
`docker: invalid reference format.`

running
`docker run -v C:\Users\rfdarter\Documents\Git\aioesphomeapi:/aioesphome ghcr.io/esphome/aioesphomeapi-proto-builder:latest`
instead runs the docker but results in this error
`docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "script/gen-protoc": stat script/gen-protoc: no such file or directory: unknown.`

I installed protoc and ran
`py script\gen-protoc`
is that fine too? Should i commit the generated C files?

**Comment by bdraco** [comment]
Here is what I get 
```
bdraco@MacBook-Pro aioesphomeapi % docker run -v "$PWD":/aioesphomeapi ghcr.io/esphome/aioesphomeapi-proto-builder:latest
bdraco@MacBook-Pro aioesphomeapi % 
```

**Comment by bdraco** [comment]
We need to use the version in the docker container so we don't introduce a protobuf code version change as well. I went ahead and ran it locally and fixed the lint issues

File: aioesphomeapi/api.proto [hunk 1]
```diff
@@ -954,6 +955,48 @@ message NumberCommandRequest {
...
+}
+message DatetimeStateResponse {
+  option (id) = 101;
+  option (source) = SOURCE_SERVER;
+  option (ifdef) = "USE_DATETIME";
+  option (no_delay) = true;
+
+  fixed32 key = 1;
+  string state = 2;
+  // If the dattime does not have a valid state yet.
```
Comment (bdraco) [inline aioesphomeapi/api.proto:986]:
```suggestion
  // If the datetime does not have a valid state yet.
```

File: tests/test_client.py [hunk 2]
```diff
@@ -617,6 +618,29 @@ async def test_number_command(
...
+            dict(key=1, state="2024-02-15 18:30:45"),
+        ),
+    ],
+)
+async def test_datetime_command(
+    auth_client: APIClient, cmd: dict[str, Any], req: dict[str, Any]
+) -> None:
+    send = patch_send(auth_client)
+
+    await auth_client.datetime_command(**cmd)
```
Comment (bdraco) [inline tests/test_client.py:640]:
```suggestion
    auth_client.datetime_command(**cmd)
```

**Comment by bdraco** [comment]
I made some breaking changes in api v22 so I adjusted your PR for them

**Review by jesserockz — APPROVED** [review]

**Pull request merged**